ath79: add support for TP-Link TL-WR802N V1 and V2

Leo-PL opened the pull request:
ath79: add support for TP-Link TL-WR802N V1 and V2

Specification:
- SoC: Qualcomm Atheros QCA9533 (560 MHz, MIPS 24Kc)
- RAM: 32 MiB
- Storage: 4 MiB of Flash on board
- Wireless: Built into QCA9533 (Honey Bee), PHY modes b/g/n
- Ethernet: 1x100M (port0)

Installation through OEM Web Interface:
- Connect to TL-WR802N by Ethernet or Wi-Fi
- Go to web interface:
  [V1] http://192.168.0.1
  [V2] http://192.168.0.254
  Default user is "admin" & password is "admin".
  On V2, there is no DHCP server running by default, so remember to set
  IP manually.
- Go to "System Tools -> Firmware Upgrade"
- Browse for firmware:
  [V1] "*.factory.bin"
  [V2] "*.factory-us.bin" or  "*.factory-eu.bin" for eu model
  Web interface may complain if filename is too long. In such case,
  rename .bin to something shorter.
- Click upgrade

Installation through tftp:
Note: T_OUT, T_IN and GND on the board must be connected to USB TTL
      Serial Configuration 115200 8n1

- Boot the TL-WR802N
- When "Autobooting in 1 seconds" appears type "tpl" followed by enter
- Connect to the board Ethernet port
    (IPADDR: 192.168.1.1, ServerIP: 192.168.1.10)
- tftpboot 0x80000000 <Firmware Image Name>
- Record the result of "printenv bootcmd"
- Enter "erase <Result of 'printenv bootcmd'> +0x3c0000"
    (e.g erase 0x9f020000 +0x3c0000)
- Enter "cp.b 0x80000000 <Result of 'printenv bootcmd'> 0x3c0000"
    (e.g cp.b 0x80000000 0x9f020000 0x3c0000)
- Enter "bootm <Result of 'printenv bootcmd'>"
    (e.g bootm 0x9f020000)

When porting from ar71xx target to ath79, I found out that on V2,
reset button is on GPIO12 and active low, instead of GPIO11 and
acticve high. By cross-flashing V1 firmware to V2, I confirmed
the same is true for V1.
Also according to manual of V1, this one also has green
LED instead of blue - both of those issues were fixed accordingly.

Installation manual based on ar71xx support by Thomas Roberts

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Compile-tested-on: TP-Link TL-WR802N v1/v2
Runtime-tested-on: TP-Link TL-WR802N v2

I have a backport for 19.07 ready and tested as well, pushed on tl-wr802n-ath79_19.07 branch on my Github.

adrianschmutzler commented on target/linux/ath79/image/tiny-tp-link.mk:
```diff
@@ -202,6 +202,30 @@ define Device/tplink_tl-wr703n
+endef
+TARGET_DEVICES += tplink_tl-wr802n-v2
+
 define Device/tplink_tl-wr740n-v1
```
Please apply proper alphabetic sorting.

adrianschmutzler commented on target/linux/ath79/image/tiny-tp-link.mk:
```diff
@@ -202,6 +202,30 @@ define Device/tplink_tl-wr703n
+  SOC := qca9533
+  DEVICE_MODEL := TL-WR802N
+  DEVICE_VARIANT := v2
+  SUPPORTED_DEVICES += tl-wr802n-v2
```
Maybe move this down to have same order of variables as for v1.

Leo-PL replied:
Done.

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_tl-wr802n.dtsi:
```diff
@@ -0,0 +1,99 @@
+
+	flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
```
I think these two lines can be omitted.

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_tl-wr802n.dtsi:
```diff
@@ -0,0 +1,99 @@
+&eth1 {
+	compatible = "syscon", "simple-mfd";
+};
+
```
Remove empty line at EOF.

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_tl-wr802n.dtsi:
```diff
@@ -0,0 +1,99 @@
+
+	phy-handle = <&swphy4>;
+
+	mtd-mac-address = <&uboot 0x1fc00>;
```
ar71xx initializes this MAC address with adding +1. Personally, I'd rather believe that your version with lan=wlan is correct.

Do you have any chance to verify the assignment with OEM firmware?

Leo-PL replied:
Yeah, just need to download it and hope that @pepe2k's u-boot_mod can boot it.

What I've found in ar71xx sources is that offset is 0, I'll also double-check that, because that source code is a bit messy.

Leo-PL replied:
You were right indeed w.r.t. vendor firmware. Adjusted accordingly.

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_tl-wr802n-v2.dts:
```diff
@@ -0,0 +1,19 @@
+
+	keys {
+		reset {
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
```
Did you evaluate this yourself? In ar71xx, the v1 variant from this PR is used for both.

Despite, I'd prefer to have the full keys node in the DTS if keys actually differ, as this is much easier to read for only one button.

Leo-PL replied:
Yes, I found this out because reset button with assignment taken from ar71xx wasn't working on V2. Once I had the system booted I could export all unassigned GPIOS and find this out by looking at /sys/kernel/debug/gpio.

Leo-PL replied:
I'll also take a look on GPL sources for V1 and compare it against V2 - this should be present somewhere there. I can do at least that not having an actual V1 to test on.

Leo-PL replied:
At least for V2 it's resolved, wonder what happens on mailing list with V1.

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_tl-wr802n.dtsi:
```diff
@@ -0,0 +1,99 @@
+		led_system: system {
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+	};
```
As for keys, I'd prefer to have the full leds node in the DTSes if it's only one LED ...

Leo-PL replied:
Moved everything back to .dtsi.

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_tl-wr802n-v1.dts:
```diff
@@ -0,0 +1,19 @@
+};
+
+&led_system {
+			label = "tp-link:blue:system";
```
On the data sheet it looks green as well ...

Leo-PL replied:
V1 manual I got from fccid.io also shows green LED, so I'll stick with it for both versions. It's possible that it was changed from blue to green in minor revision of V1.

Leo-PL replied:
Unified both of them

adrianschmutzler commented on target/linux/ath79/tiny/base-files/etc/board.d/01_leds:
```diff
@@ -96,6 +96,12 @@ tplink,tl-wr941-v4)
 	ucidef_set_led_switch "lan4" "LAN4" "tp-link:green:lan4" "switch0" "0x10"
 	;;
+tplink,tl-wr802n-v1)
+	ucidef_set_led_wlan "wlan" "WLAN" "tp-link:blue:system" "phy0tpt"
```
Is this vendor behavior? Please use a DTS trigger if that's possible with the diag LEDs ...

Leo-PL replied:
It's more complicated than that. Manual says that this is supposed to be for WLAN activity, but the actual behaviour I got out of box is different - the light would be solid when Ethernet link is established, and blink otherwise.

So here, I'd rather be consistent with rest of OpenWrt than manufacturer firmware, and drop this trigger altogether, and use this LED as named, for system status.

adrianschmutzler replied:
Would be fine for me as well.

Leo-PL replied:
OK, dropped.

adrianschmutzler commented on target/linux/ath79/tiny/base-files/etc/board.d/02_network:
```diff
@@ -43,7 +43,9 @@ ath79_setup_interfaces()
-	tplink,tl-wr703n)
+	tplink,tl-wr703n|\
+	tplink,tl-wr802n-v1|\
+	tplink,tl-wr802n-v2|)
```
remove the pipe ...

adrianschmutzler commented on target/linux/ath79/tiny/base-files/etc/uci-defaults/04_led_migration:
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
```
please remove the shebang, these files are sourced.

Leo-PL replied:
Actually, this is a copy-paste leftover from generic subtarget, so it's need to be fixed there also. Will fix here of course :)

adrianschmutzler replied:
I'm aware, should be removed there as well. Guess what I compared this with for review ...

adrianschmutzler commented on target/linux/ath79/tiny/base-files/etc/uci-defaults/04_led_migration:
```diff
@@ -0,0 +1,17 @@
+
+case "$board" in
+tplink,tl-wr802n-v2)
+	migrate_leds "^$boardonly:=tp-link:"
```
This is correct at the moment, but the whole file would become obsolete if the DTS trigger works.

Leo-PL replied:
If we stick to not defining trigger altogether (see above), this still has merit to migrate any configuration users might have made previously.

adrianschmutzler replied:
Fair point. Didn't think of that.
The only question is whether this also applies for a ar71xx->ath79 migration, where the user is encouraged to upgrade without keeping settings anyway. So, in general I'd follow your argument, but for this specific case I think the user should be "made aware" that something changed (i.e. the "recommended" LED behavior), and given the chance to start from the "new default" setup.

However, this is just for your consideration, if you prefer to keep the led_migration I will be fine with that as well.

Leo-PL replied:
I think it doesn't hurt to have this, and personally wouldn't like this kind of encouragement. I guess that a lot of people will just forget about setting this up anyway :)

adrianschmutzler commented:
> While testing I found out that at least for V2 the GPIO configuration for reset is different than what was found in ar71xx - GPIO12 active low. So I wonder if this isn't also the case for v1 - I would really appreciate a test for this one case. Before that discovery I thought that they differ only by SoC revision (AL1A vs BL1A) and LED colour (blue vs green).

That's why this should be added (rephrased) to the commit message. Then, I would have read it before writing the comments.

Based on the differences and uncertainties, I would only merge v2 unless we can find a tester for v1.

> I have a backport for 19.07 ready and tested as well, pushed on tl-wr802n-ath79_19.07 branch on my Github.

I have not yet made up my mind whether I will backport this. In any case, I'd do the backport myself anyway, as this is effectively less work that reviewing somebody else's backport :-)
I'd push this to separate staging for testing then, but I don't guarantee anything.

adrianschmutzler commented:
@schuza @p-wassi Does one of you still have a TL-WR802N v1 and could verify the reset GPIO (11 ACTIVE_HIGH) and LED color (blue) on that device?

adrianschmutzler commented:
I just had a look into the manual. Based on what's described there, I think we should not use an activity trigger like phy0tpt, as blinking of that led would mean it's booting. Instead, we should have something like phy0radio (is this available here?) which gives a solid light for WiFi, so blinking is clearly an indication of boot/failsafe/whatever ...

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_tl-wr802n.dtsi:
```diff
@@ -0,0 +1,99 @@
+	aliases {
+		led-boot = &led_system;
+		led-failsafe = &led_system;
+		led-running = &led_system;
```
Since the LED should be used for WiFi, drop led-running. By that, it won't stay on after successful boot, which might be mistaken for enabled wifi ...

Leo-PL replied:
> 
> 
> > While testing I found out that at least for V2 the GPIO configuration for reset is different than what was found in ar71xx - GPIO12 active low. So I wonder if this isn't also the case for v1 - I would really appreciate a test for this one case. Before that discovery I thought that they differ only by SoC revision (AL1A vs BL1A) and LED colour (blue vs green).
> 
> That's why this should be added (rephrased) to the commit message. Then, I would have read it before writing the comments. ;-)
> 
> This should also be fixed for ar71xx in (a) separate commit(s).
> 
> Based on the differences and uncertainties, I would only merge v2 unless we can find a tester for v1.
> 
> > I have a backport for 19.07 ready and tested as well, pushed on tl-wr802n-ath79_19.07 branch on my Github.
> 
> I have not yet made up my mind whether I will backport this. In any case, I'd do the backport myself anyway, as this is effectively less work that reviewing somebody else's backport :-)
> I'd push this to separate staging for testing then, but I don't guarantee anything.

Sure, will do. Fixes for ar71xx are coming too (separately), at least for consistency.

Regarding backport - full understanding, given that it took me about 5 minutes to do this myself, having done support on master. Which was a kind of a 'porting speedrun' clocking in at 03:09 start to finish ;)

Leo-PL replied:
Also see my comment on the trigger in 01_leds for this one :)

adrianschmutzler replied:
Well, if the LED is really just a system LED, it would be better to have led-running. Having it off would look strange (like something is broken).

However, if somebody customizes the behavior, the led-running will be a annoying.

I'll leave it to you, so I don't have to decide :-)

Leo-PL replied:
Actually I haven't had much problems if led-running was present and an override was made in /etc/config/system. The LED just goes on to steady state for a while before new trigger is set by UCI. So let's keep it, as it is most consistent with other devices supported by OpenWrt.

Leo-PL commented:
> 
> 
> > While testing I found out that at least for V2 the GPIO configuration for reset is different than what was found in ar71xx - GPIO12 active low. So I wonder if this isn't also the case for v1 - I would really appreciate a test for this one case. Before that discovery I thought that they differ only by SoC revision (AL1A vs BL1A) and LED colour (blue vs green).
> 
> That's why this should be added (rephrased) to the commit message. Then, I would have read it before writing the comments. ;-)
> 
> This should also be fixed for ar71xx in (a) separate commit(s).
> 
> Based on the differences and uncertainties, I would only merge v2 unless we can find a tester for v1.
> 
> > I have a backport for 19.07 ready and tested as well, pushed on tl-wr802n-ath79_19.07 branch on my Github.
> 
> I have not yet made up my mind whether I will backport this. In any case, I'd do the backport myself anyway, as this is effectively less work that reviewing somebody else's backport :-)
> I'd push this to separate staging for testing then, but I don't guarantee anything.

Sure, will do. Fixes for ar71xx are coming too (separately), at least for consistency.

Regarding backport - full understanding, given that it took me about 5 minutes to do this myself, having done support on master. Which was a kind of a 'porting speedrun' clocking in at 03:09 start to finish ;)

Leo-PL commented:
My initial tinkering with manufacturers GPL source seems to lead to believe that GPIO12 is indeed used for reset, but TP-Link didn't attach proper kernel config for V1. In sources for V2, there is clear indication that reset is at GPIO12 directly in provided config. V1 only shows so after running `make board953x_defconfig`. Kernel source tree seems to be _very_ similar.

adrianschmutzler commented:
@Leo-PL I also tend to believe that your findings will apply to both v2 and v1.

I don't really like the idea of merging v1 with a real on-device test. However, based on the recent development I'd recommend to create a patch for v2 only, which already contains a DTSI based on our recent findings (i.e. LEDs and keys in the DTSI). After this been merged, you could open a second trivial PR for v1, and then we see what will happen to that one. This has the advantage that most of the review burden etc. is already done, and we could continue quickly if there is a tester for v1.
For ar71xx, it's less fun. We could split the buttons for now, but I don't think that's really desirable. I wouldn't touch the LED labels there at all.

adrianschmutzler commented:
Note this forum comment:
https://forum.archive.openwrt.org/viewtopic.php?id=46475&p=2
(just text-search for "reset", cannot link the post itself).

Seems like the button doesn't work on v1 indeed.

Leo-PL commented:
I've cross-flashed stock firmware of V1 on my V2, and guess what...
Everything worked identically to V2 firmware, reset button included.

pepe2k's  u-boot_mod also reacts correctly to the reset button on GPIO12 while stating support only for V1.

In GPL sources, reset button is bound only to one GPIO at a time through `request_irq`.Those three findings give me 100% confidence that mapping is the same. Probably whole hardware too. I've seen some shots of V1.1 from DD-WRT forum showing QCA9533-BL3A as in V2:
https://forum.dd-wrt.com/phpBB2/viewtopic.php?p=1125424
It's likely that only difference between V1 and V2 is more strict region validation for firmware, which was the cause for differentiating between EU and US firmwares.

I also checked MAC offsets and you were right - label is at WLAN, and Ethernet is +1 from that.

I'll push updated PR very soon :-)

Leo-PL committed:
ath79: add support for TP-Link TL-WR802N V1 and V2

Specification:
- SoC: Qualcomm Atheros QCA9533 (560 MHz, MIPS 24Kc)
- RAM: 32 MiB
- Storage: 4 MiB of Flash on board
- Wireless: Built into QCA9533 (Honey Bee), PHY modes b/g/n
- Ethernet: 1x100M (port0)

Installation through OEM Web Interface:
- Connect to TL-WR802N by Ethernet or Wi-Fi
- Go to web interface:
  [V1] http://192.168.0.1
  [V2] http://192.168.0.254
  Default user is "admin" & password is "admin".
  On V2, there is no DHCP server running by default, so remember to set
  IP manually.
- Go to "System Tools -> Firmware Upgrade"
- Browse for firmware:
  [V1] "*.factory.bin"
  [V2] "*.factory-us.bin" or  "*.factory-eu.bin" for eu model
  Web interface may complain if filename is too long. In such case,
  rename .bin to something shorter.
- Click upgrade

Installation through tftp:
Note: T_OUT, T_IN and GND on the board must be connected to USB TTL
      Serial Configuration 115200 8n1

- Boot the TL-WR802N
- When "Autobooting in 1 seconds" appears type "tpl" followed by enter
- Connect to the board Ethernet port
    (IPADDR: 192.168.1.1, ServerIP: 192.168.1.10)
- tftpboot 0x80000000 <Firmware Image Name>
- Record the result of "printenv bootcmd"
- Enter "erase <Result of 'printenv bootcmd'> +0x3c0000"
    (e.g erase 0x9f020000 +0x3c0000)
- Enter "cp.b 0x80000000 <Result of 'printenv bootcmd'> 0x3c0000"
    (e.g cp.b 0x80000000 0x9f020000 0x3c0000)
- Enter "bootm <Result of 'printenv bootcmd'>"
    (e.g bootm 0x9f020000)

When porting from ar71xx target to ath79, I found out that on V2,
reset button is on GPIO12 and active low, instead of GPIO11 and
acticve high. By cross-flashing V1 firmware to V2, I confirmed
the same is true for V1.
Also according to manual of V1, this one also has green
LED instead of blue - both of those issues were fixed accordingly.

Installation manual based on ar71xx support by Thomas Roberts

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_tl-wr802n.dtsi:
```diff
@@ -0,0 +1,99 @@
+	phy-handle = <&swphy4>;
+
+	mtd-mac-address = <&uboot 0x1fc00>;
+	mtd-mac-address-increment = <1>;
```
Okay, so ar71xx was right about that. Actually, I would have expected eth0 to be the same as wifi ... Seems like TP-Link has thrown the dice differently here.

Leo-PL replied:
Indeed it's quite surprising.

adrianschmutzler commented:
Okay, so the combination of ar71xx code, your cross-flashing and the forum report I found convinces me that v1 is most probably right, unless somebody invalidates this responding to list.

So, I think I will merge this as-is in a few days.

Can you please prepare a separate commit for ar71xx (based on master as well), I think the only thing left to fix there would be the button (for both devices based on the same justification you gave in the recent commit message).
Just push the ar71xx commit to this PR as well ...

Leo-PL committed:
ar71xx: fix GPIO_RESET for TP-Link TL-WR802N V1/V2

During porting support for this router to ath79 target,
it was discovered that GPIO mapping was incorrect (GPIO11 active high).
Correct mapping for both V1 and V2 is GPIO12 active low.

Default configuration from GPL source for V2 explicitly states this, and
this was confirmed experimentally on ath79 by looking on
/sys/kernel/debug/gpio. Correctness of this was also validated for V1 by
cross-flashing vendor firmware for V1 on V2 hardware, in which reset
button also worked.

Fix it.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL commented:
OK, fix for ar71xx in bf72c85. Tested as well.

I don't want to touch the LEDs in ar71xx either.

adrianschmutzler commented:
Thanks, I've pulled this into my staging tree. I will post a link to a backported version here later.

Leo-PL commented:
OK, thank you too :-)

adrianschmutzler commented:
I've pushed a backport here. Please review and test if possible:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging-19.07

Leo-PL commented:
Since you ended up with almost line-to-line identical backport as I did, save for ordering of two variables in makefile, and it already works for me: +1.

adrianschmutzler commented:
Thanks, backported as well, closing now.
